skypy/galaxies/velocity_dispersion.py

```diff
@@ -0,0 +1,69 @@
+r"""Sample from velocity distribution function
+
+"""
+
+import numpy as np
+import scipy.special as special
+
+
+def schecter_vdf(vd_min, vd_max, resolution=100, size=None, scale=1.):
+    r"""Sample from velocity dispersion function of elliptical galaxies in the local universe [1]_.
+
+    Parameters
+    ----------
+    vd_min, vd_max: int
+        Lower and upper bounds of random variable x. Samples are drawn uniformly from bounds.
+    resolution: int
+        Resolution of the inverse transform sampling spline. Default is 100.
+    size: int
+        Number of samples returned. Default is 1.
+
+    Returns
+    -------
+    velocity_dispersion: array_like
+        Velocity dispersion drawn from Schechter function.
+
+    Warnings
+    --------
+    Inverse cumulative dispersion function is approximated from the function
+    using quadratic interpolation. The user should specify the resolution to
+    satisfy their numerical accuracy.
+
+        Notes
+    -----
+    The probability distribution function :math:`p(\sigma)` for redshift :math:`\sigma`
+    is given by Choi et al. [1]_ as
+
+    .. math::
+        \begin{eqnarray}&&{dn}={\phi }_{*}{(\displaystyle \frac{\sigma }{{\sigma }_{*}})}^{\alpha }
+        \mathrm{exp}[-{(\displaystyle \frac{\sigma }{{\sigma }_{*}})}^{\beta }]\displaystyle
+        \frac{\beta }{{\rm{\Gamma }}(\alpha /\beta )}\displaystyle \frac{1}{\sigma },
+        \end{eqnarray}\tag{3} \;.
+
+    References
+    ----------
+    .. [1] Choi, Park and Vogeley, (2007), astro-ph/0611607, doi:10.1086/511060
+
+    """
+
+    if size is None:
+        size = np.broadcast(vd_min, vd_max, scale).shape or None
+
+    lnx = np.linspace(vd_min, vd_max, resolution)
+
+    def vdf_func(x):
+        return 8e-3*(x/161)**2.32*np.exp(-(x/161)**2.67)*(2.67/special.gamma(2.32/2.67))*(1/x)
+
+    pdf = vdf_func(lnx)
+    cdf = pdf  # in place
+    np.cumsum((pdf[1:]+pdf[:-1])/2*np.diff(lnx), out=cdf[1:])
+    cdf[0] = 0
+    cdf /= cdf[-1]
+
+    t_lower = np.interp(vd_min, lnx, cdf)
+    t_upper = np.interp(vd_max, lnx, cdf)
+
+    u = np.random.uniform(t_lower, t_upper, size=size)
+    lnx_sample = np.interp(u, cdf, lnx)
+
+    return lnx_sample * scale
```

Comment:
Hi, Philipp. I am not entirely sure how to do this. The velocity dispersion Schechter function is a modified version of the normal Schechter function, so it has a part which is `\exp[-{(\frac{\sigma}{{\sigma}_{*}})}^{\beta }]`, which means there is a beta parameter raised inside the exponent, which I think skypy's schechter function does not support. Perhaps I am misunderstanding.

 Also, in terms of adding an option for setting alpha and beta, do I leave the default as the values calculated in Choi et al.? E.g. `def schechter_vdf(alpha=2.32, ....`.

Reply:
To the first:
This is not a problem. You can re-parametrise and bring it into the same form since beta is a constant. If you start from equation 4 from the paper,
$$\phi(\sigma)\mathrm{d}\sigma = \phi_\* (\sigma/\sigma_\*)^\alpha \cdot e^2$$

Reply:
To the second:
No don't include the defaults of the Choi paper.

Reply:
Ah you are right with the exponent. This was my mistake. Sorry and thanks.
Your approach was already correct. So if we start with 
$$\phi(\sigma) \mathrm{d}\sigma = \phi_\* \left(\frac{\sigma}{\sigma_\*}\right)^\alpha \exp\left[-\left( \frac{\sigma}{\sigma_\*} \right)^\beta\right] \frac{\beta}{\Sigma(\alpha/\beta)} \frac{1}{\sigma} \mathrm{d}\sigma,$$
we can define
$$x = \left( \frac{\sigma}{\sigma_\*} \right)^\beta.$$
Taking the derivative we get
$$\frac{\mathrm{d}x}{\mathrm{d}\sigma} = \beta x \frac{1}{\sigma}$$
and
$$\mathrm{d}\sigma = \frac{\sigma}{\beta x}\mathrm{d}x.$$
Plugging this into the first equation we have
$$\phi(x)\mathrm{d}x = \phi_\* \left(\frac{\sigma}{\sigma_\*}\right)^\alpha \exp[-x] \frac{\beta}{\Sigma(\alpha/\beta)} \frac{1}{\sigma} \frac{\sigma}{\beta x} \mathrm{d}x = \frac{\phi_\*}{\Sigma(\alpha/\beta)} \frac{1}{x} \left(x^{\frac{1}{\beta}}\right)^\alpha \exp[-x] \mathrm{d}x = \frac{\phi_\*}{\Sigma(\alpha/\beta)} x^{\frac{\alpha}{\beta} - 1} \exp[-x] \mathrm{d}x.$$
Now we can define $\alpha'=\alpha/\beta - 1$ and get
$$\phi(x) = \frac{\phi_\*}{\Sigma(\alpha/\beta)} x^{\alpha'} \exp[-x].$$

I hope I didn't do any other mistake. Please correct me if this is the case.

Reply:
Thank you very much. I have implemented what I think you mean, but I am not sure if it is right. To test it against my own sampling, how do I normalise my function? Also, I have omitted the constant

$$\frac{\phi_*}{\Sigma(\alpha/\beta)}$$

Is this the correct thing?

Reply:
Thanks. Omitting the normalisation is fine in this case because in `random.schechter()` we calculate the cdf as a cumulative sum of the pdf and normalise it afterwards. 

To test if everything is correct we have the unit tests. You can do an independent sampling for a set of parameters and compare it to the sampling with this implementation. For this, you can use the implementation that you had initially. That would be a very good test. However, I took a first quick look at the new commits. Pay attention to the boundaries `vd_min` and `vd_max`. You need to change them to an `x_min` and `x_max`.

If you could add the unit test (you can find an example [here](https://github.com/Vikramb1/skypy/blob/df2536974c122dd72498f8d146913c14c832833b/skypy/galaxies/tests/test_stellar_mass.py#L53)) I can take another look early next week.

Reply:
Sure. I have a quick question about testing sampling. I have created a package some time ago: https://github.com/Vikramb1/Lens_population_sampler which I can import and test sampling from, but I would assume you would prefer the implementation without the importing of other packages, right?

Reply:
Yes, without other packages please